zc_common/remote_resource/filters.py

```diff
@@ -1,55 +1,119 @@
+    from rest_framework import filterset
 except ImportError:
     from django_filters.rest_framework import DjangoFilterBackend
+    from django_filters.rest_framework.filters import Filter
+    from django_filters.rest_framework import filterset
+
+
+class ArrayFilter(Filter):
+    field_class = SimpleArrayField
+
+    @property
+    def field(self):
+        # This property needs to be overriden because filters.Filter does not instantiate field_class with any
+        # args by default, and SimpleArrayField requires an argument indicating the type of each element in the array
+        self._field = self.field_class(forms.CharField(), required=False)
+        return self._field
+
+
+class JSONAPIFilterSet(filterset.FilterSet):
+    class Meta:
+        strict = True
+        filter_overrides = {
+            ArrayField: {
+                'filter_class': ArrayFilter,
+                'extra': lambda f: {
+                    'lookup_expr': 'contains',
+                }
+            }
+        }
 
 
 class JSONAPIFilterBackend(DjangoFilterBackend):
+    default_filter_set = JSONAPIFilterSet
+
+    # This method takes the filter query string (looks something like ?filter[xxx]=yyy) and parses into parameters
+    # that django_filters can interface with.
+    #
+    # Handles:
+    #   ?filter[id]=1
+    #   ?filter[id__in]=1,2,3
+    #   ?filter[price__gte]=100
+    #   ?filter[relatedobject__relatedobject]=1
+    #   ?filter[relatedobject__relatedobject__in]=1,2,3
+    def _parse_filter_string(self, queryset, filter_class, filter_string, filter_value):
+        filter_string_parts = filter_string.split('__')
+        if len(filter_string_parts) > 1:
+            field_name = '__'.join(filter_string_parts[:-1])
+        else:
+            field_name = filter_string_parts[0]
+
+        filterset_data = []
+
+        # Translates the 'id' in ?filter[id]= into the primary key identifier, e.g. 'pk'
+        if field_name == 'id':
+            primary_key = queryset.model._meta.pk.name
+            field_name = primary_key
+
+        try:
+            is_many_to_many_field = isinstance(getattr(queryset.model, filter_string).field, ManyToManyField)
+            if is_many_to_many_field:
+                filter_value = filter_value.split(',')
+        except AttributeError:
+            pass
+
+        try:
+            field_name = filter_string.split('__')[0]
+            field_filter = filter_class.get_filters().get(field_name, None)
+            is_array_filter = isinstance(field_filter, ArrayFilter)
+            if is_array_filter:
+                filter_value = filter_value.split(',')
+        except AttributeError:
+            pass
+
+        # Allow 'true' or 'false' as values for boolean fields
+        try:
+            if isinstance(queryset.model._meta.get_field(field_name), BooleanField):
+                filter_value = bool(strtobool(filter_value))
+        except FieldDoesNotExist:
+            pass
+
+        filterset_data.append({
```

Comment:
It seems like there would never be more than one element in this array since we don't do any looping within this function, would it be better to return the dict itself or is there a reason we're keeping it as a list?